crypto/store/store_register.c

```diff
@@ -0,0 +1,226 @@
+}
+
+static int store_loader_cmp(const STORE_LOADER *a, const STORE_LOADER *b)
+{
+    if (a->scheme != NULL && b->scheme != NULL)
+        return strcmp(a->scheme, b->scheme);
+    else if (a->scheme == b->scheme)
+        return 0;
+    return a->scheme == NULL ? -1 : 1;
+}
+
+static LHASH_OF(STORE_LOADER) *loader_register = NULL;
+
+int store_register_loader_int(STORE_LOADER *loader)
+{
+    const char *scheme = loader->scheme;
+
+    /*
+     * Check that the given scheme conforms to correct scheme syntax as per
+     * RFC 3986:
+     *
+     * scheme        = ALPHA *( ALPHA / DIGIT / "+" / "-" / "." )
+     */
+    if (isalpha(*scheme))
+        while (*scheme != '\0'
+               && (isalpha(*scheme)
+                   || isdigit(*scheme)
+                   || strchr("+-.", *scheme) != NULL))
+            scheme++;
+    if (*scheme != '\0') {
+        STOREerr(STORE_F_STORE_REGISTER_LOADER_INT, STORE_R_INVALID_SCHEME);
+        ERR_add_error_data(4, "scheme=", loader->scheme);
+        return 0;
+    }
+
+    if (loader_register == NULL) {
+        loader_register = lh_STORE_LOADER_new(store_loader_hash,
+                                              store_loader_cmp);
+        if (loader_register == NULL)
+            return 0;
+    }
+
+    if (lh_STORE_LOADER_insert(loader_register, loader) == NULL
+        && lh_STORE_LOADER_error(loader_register) > 0)
+        return 0;
+
+    return 1;
+}
+int STORE_register_loader(STORE_LOADER *loader)
+{
+    if (!store_init_once())
+        return 0;
+    return store_register_loader_int(loader);
+}
+
+const STORE_LOADER *store_get0_loader_int(const char *scheme)
+{
+    STORE_LOADER template;
+    STORE_LOADER *loader = NULL;
+
+    template.scheme = scheme;
+    template.open = NULL;
+    template.load = NULL;
+    template.eof = NULL;
+    template.close = NULL;
+
+    if (!store_init_once())
+        return 0;
+
+    loader = lh_STORE_LOADER_retrieve(loader_register, &template);
+
+    if (loader == NULL) {
+        STOREerr(STORE_F_STORE_GET0_LOADER_INT, STORE_R_UNREGISTERED_SCHEME);
+        ERR_add_error_data(2, "scheme=", scheme);
+        return 0;
+    }
+
+    return loader;
+}
+
+STORE_LOADER *store_unregister_loader_int(const char *scheme)
+{
+    STORE_LOADER template;
+    STORE_LOADER *loader = NULL;
+
+    template.scheme = scheme;
+    template.open = NULL;
+    template.load = NULL;
+    template.eof = NULL;
+    template.close = NULL;
+
+    loader = lh_STORE_LOADER_delete(loader_register, &template);
+
+    if (loader == NULL) {
+        STOREerr(STORE_F_STORE_UNREGISTER_LOADER_INT,
+                 STORE_R_UNREGISTERED_SCHEME);
+        ERR_add_error_data(2, "scheme=", scheme);
+        return 0;
+    }
+
+    return loader;
+}
+STORE_LOADER *STORE_unregister_loader(const char *scheme)
+{
+    if (!store_init_once())
+        return 0;
+    return store_unregister_loader_int(scheme);
+}
+
+void destroy_loaders_int(void)
+{
+    lh_STORE_LOADER_free(loader_register);
```

Comment:
Shouldn't be needed.  If you look in `crypto/init.c`, you'll notice that `engine_cleanup_int` is called before `store_cleanup_int`, meaning the engines should have unregistered their store loaders.

Maybe I should add an extra check here that all loaders have been unregistered (checking that there are zero items in `loader_register`) and cry bloody murder (what would be the best way to do that?  `OPENSSL_assert`?  I know you'd love that ;-) ) is not...